### DISCUSSION TITLE
aqua update-aqua doesn't work on Git Bash: `The process cannot access the file because it is being used by another process`

### DISCUSSION THREAD

**Discussion by suzuki-shunsuke** [aquaproj/aqua#2510] [Deprecated (Bug Report)] [answered]
### aqua info

```console
$ aqua info
{
"version": "2.16.0",
"commit_hash": "70f84d66ba9af98008360fd7b846094f020b4c6f",
"os": "windows",
"arch": "amd64",
"pwd": "C:\Users\devnu",
"root_dir": "C:\Users\devnu\AppData\Local\aquaproj-aqua",
"env": {},
"config_files": []
}
```

### Overview

`aqua update-aqua` fails on Git Bash.

### How to reproduce

Executed command and output

```console
$ aqua update-aqua
```

### Expected behaviour

`aqua update-aqua` succeeds and aqua is updated.

### Actual behaviour

`aqua update-aqua` failed.

```console
$ aqua upa v2.16.1
FATA[0000] aqua failed aqua_version=2.16.0 env=windows/amd64 error="download aqua: create a file: open C:\Users\devnu\AppData\Local\aquaproj-aqua\bin\aqua.exe: The process cannot access the file because it is being used by another process." new_version=v2.16.1 program=aqua
```

**Comment by suzuki-shunsuke** [comment]
```
download aqua:
create a file:
open C:\Users\devnu\AppData\Local\aquaproj-aqua\bin\aqua.exe:
The process cannot access the file because it is being used by another process.
```

https://github.com/aquaproj/aqua/blob/8fa422b58b9b2142129b4fcc254ed41ae02478c9/pkg/controller/updateaqua/controller.go#L48-L52

https://github.com/aquaproj/aqua/blob/8fa422b58b9b2142129b4fcc254ed41ae02478c9/pkg/installpackage/aqua.go#L135-L137

https://github.com/aquaproj/aqua/blob/8fa422b58b9b2142129b4fcc254ed41ae02478c9/pkg/installpackage/copy.go#L14-L17

**Reply by suzuki-shunsuke** [reply]
On Windows, symbolic links don't work well so aqua don't use symbolic links on Windows.
So `aqua update-aqua` tries to replace the executable file `aqua.exe` with a new version, but it fails because `aqua.exe` is being executed.

I have no idea to resolve the issue for now.

**Reply by suzuki-shunsuke** [reply]
One idea is to create a tool to manage aqua like `aquaenv`.

**Reply by CrystalMethod** [reply]
A side note: when using the Scoop package manager on Windows, the command `aqua update-aqua` does not work correctly either. However, the update itself runs successfully, i.e. the new version has been downloaded. But, since Scoop also uses shims, the command `aqua` is not directed to the new version, but still to the version installed by Scoop. (by design)

```console
$ aqua update-aqua v2.18.0
INFO[0000] download and unarchive the package            aqua_version=2.19.0 env=windows/arm64 new_version=v2.18.0 package_name=aquaproj/aqua package_version=v2.18.0 program=aqua registry=
INFO[0000] verify a package with slsa-verifier           aqua_version=2.19.0 env=windows/arm64 new_version=v2.18.0 package_name=aquaproj/aqua package_version=v2.18.0 program=aqua registry=
Verified signature against tlog entry index 50787879 at URL: https://rekor.sigstore.dev/api/v1/log/entries/24296fb24b8ad77a6423ecfb6e7b9f7af6564f22f6d7c0ad9b8d18f4f2f89ff60c267dd238ddb44d
Verified build using builder "https://github.com/slsa-framework/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@refs/tags/v1.9.0" at commit 3d3b4f3e55ec7db75f7c90b55c19c9bbd2bf430b
Verifying artifact C:\Users\...\AppData\Local\Temp\189897499: PASSED

PASSED: Verified SLSA provenance

$ aqua -v
aqua version 2.19.0 (8fa422b58b9b2142129b4fcc254ed41ae02478c9)
```

So you have to use `scoop update` in this case
```console
$ scoop update aqua
aqua: 2.19.0 (latest version)
Latest versions for all apps are installed! For more information try 'scoop status'
```
But you can only update to the latest release, no downgrade is possible

```console
$ scoop install aqua@2.18.0
WARN  Given version (2.18.0) does not match manifest (2.19.0)
WARN  Attempting to generate manifest for 'aqua' (2.18.0)
'aqua' does not have autoupdate capability
couldn't find manifest for 'aqua@2.18.0'
```

But I think this is doable with some enhancements to `aquaproj/scoop-bucket`.

**Comment by suzuki-shunsuke** [comment]
There are several solutions.
We'll try them.

https://stackoverflow.com/questions/1211948/best-method-for-implementing-self-updating-software

**Reply by suzuki-shunsuke** [reply]
- https://github.com/aquaproj/aqua/pull/2871

**Reply by suzuki-shunsuke** [reply, marked as answer]
This issue has been solved by v2.27.4.
- https://github.com/aquaproj/aqua/pull/2871
- https://github.com/aquaproj/aqua/releases/tag/v2.27.4
